typing/typecore.ml

```diff
@@ -1959,15 +1998,7 @@ let type_pattern_list
   =
   reset_pattern allow;
   let new_env = ref env in
-  let type_pat (attrs, pat) ty =
-    let stack_annot = has_stack_annot attrs in
-    let mode =
-      match stack_annot, no_existentials with
-      | true, At_toplevel ->
-         raise (Error(pat.ppat_loc, env, Local_value_escapes))
```

Comment:
Where has this check gone? The point is to ensure that module-level bindings are not inferred as "local".

Reply:
Rather than insisting that all top-level bindings are heap, we check that things are heap at the point that they are added to a struct. It's not much harder to do the check that way and it lines up more closely with the semantics. It is also more expressive although not in an especially useful way:
```ocaml
let { contents = x } = stack_ ref 5
```